### ISSUE TITLE
2.10: dataclass config issues (`arbitrary_types_allowed`, aliases, etc)

### ISSUE THREAD

**Issue by jchalupka-pan** [pydantic/pydantic#10917]
> Edited by @Viicos, see below for the original issue

Pydantic <2.10 had a bug with dataclasses, where the config of the model that uses the dataclass as an annotation would be used.

This means that the following would work:

```python
from dataclasses import dataclass

from pydantic import BaseModel, ConfigDict

class ArbitraryType: pass

@dataclass
class DC:
    a: ArbitraryType


class Model(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    dc: DC
```

But it was inconsistent for two reasons:

- if `DC` were to be a Pydantic model, it wouldn't inherit the config of `Model`
- if you happen to set an unrelated configuration value on `DC` (as per [the docs](https://docs.pydantic.dev/latest/concepts/config/#configuration-with-dataclass-from-the-standard-library-or-typeddict)), it wouldn't inherit the `arbitrary_types_allowed` configuration value from the parent class.

This was fixed in https://github.com/pydantic/pydantic/pull/10576.

However, we realize it might be hard to set a configuration on `DC`, as it is a stdlib dataclasses and thus, in some scenarios, might be a type from an external library/application that users do not control.

Use cases welcomed, outcome of this issue still TBD.

<details>

<summary>Initial issue description (hidden for clarity):</summary>

### Description

Pydantic cannot generate core schema (and therefore create the model class) for model with `arbitrary_types_allowed=True` and a dataclass field that contains "arbitrary type".

See the example code below. `WorkingModel` is created correctly, the `NotWorkingModel` is not, exception is raised:

```raw
pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class '__main__.ArbitraryClass'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.

If you got this error by calling handler(<some type>) within `__get_pydantic_core_schema__` then you likely need to call `handler.generate_schema(<some type>)` since we do not call `__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.

For further information visit https://errors.pydantic.dev/2.10/u/schema-for-unknown-type
```

The workaround is to define dataclass like this:

```python
@dataclass
class WrappingDataClass:
    vsm_object: ArbitraryClass
    __pydantic_config__ = ConfigDict(arbitrary_types_allowed=True)
```

### Example Code

```Python
from dataclasses import dataclass

from pydantic import BaseModel, ConfigDict


class ArbitraryClass: ...


@dataclass
class WrappingDataClass:
    vsm_object: ArbitraryClass


class WorkingModel(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    my_field: ArbitraryClass


class NotWorkingModel(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    my_field: WrappingDataClass
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.10.0
        pydantic-core version: 2.27.0
          pydantic-core build: profile=release pgo=false
                 install path: C:\repos\panser\panser-fortigate\.venv\Lib\site-packages\pydantic
               python version: 3.11.9 (tags/v3.11.9:de54cf5, Apr  2 2024, 10:12:12) [MSC v.1938 64 bit (AMD64)]
                     platform: Windows-10-10.0.22631-SP0
             related packages: mypy-1.12.1 typing_extensions-4.12.2
                       commit: unknown
```

</details>

**Comment by bschoenmaeckers** [comment]
We used this functionality to create a Fastapi response model created from a dataclass defined in a external dependency.
The external dataclass also contains nested dataclasses so it would be a pain to override all internal dataclasses with a manual crafted pydantic model.

https://github.com/duo-labs/py_webauthn/blob/3504cd706204ffc67371e485126051cfaed5fb0c/webauthn/helpers/structs.py#L284-L308

```python
from webauthn.helpers.structs import PublicKeyCredentialCreationOptions
from pydantic import BaseModel

class PasskeyRegistrationChallenge(BaseModel, PublicKeyCredentialCreationOptions):
    model_config = {
        "ser_json_bytes": "base64",
        "alias_generator": snake_case_to_camel_case,
        "populate_by_name": True,
    }
```

**Comment by sydney-runkle** [comment]
Upon further consideration, we've decided we definitely should fix this.

Pydantic types have strict config boundaries (models, pydantic dataclasses, etc), but folks are relying on config pushdown for nested vanilla dataclasses and typed dicts, so we'll fix this promptly for v2.10.1.

**Comment by sydney-runkle** [comment]
Hey folks! We've just released [v2.10.1](https://github.com/pydantic/pydantic/releases/tag/v2.10.1) with a fix for this issue!

Let us know if you're still experiencing any difficulties. Thanks!
